src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3070,380 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+	}
+
+	/*
+	 * If stopOnFailure is true, we just error out and code execution should
+	 * never reach to this point. This is the case for reference tables and
+	 * copy from worker nodes.
+	 */
+	Assert(!stopOnFailure || failedPlacementCount == 0);
+
+	MemoryContextReset(localContext);
+}
+
+
+/*
+ * CopyGetPlacementConnection assigns a connection to the given placement. If
+ * a connection has already been assigned the placement in the current transaction
+ * then it reuses the connection. Otherwise, it requests a connection for placement.
+ */
+static MultiConnection *
+CopyGetPlacementConnection(ShardPlacement *placement, HTAB *connectionStateHash,
+						   bool stopOnFailure)
+{
+	MultiConnection *connection = NULL;
+	uint32 connectionFlags = FOR_DML;
+	char *nodeUser = CurrentUserName();
+	ShardPlacementAccess *placementAccess = NULL;
+
+	/*
+	 * Determine whether the task has to be assigned to a particular connection
+	 * due to a preceding access to the placement in the same transaction.
+	 */
+	placementAccess = CreatePlacementAccess(placement, PLACEMENT_ACCESS_DML);
+	connection = GetPlacementListConnectionIfCached(connectionFlags,
+													list_make1(placementAccess),
+													NULL);
+	if (connection != NULL)
+	{
+		return connection;
+	}
+
+	/* if execution mode set to sequential, reuse worker connection if available */
+	if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+	{
+		connection = GetWorkerConnectionIfCached(placement->nodeName,
+												 placement->nodePort,
+												 connectionStateHash);
+		if (connection != NULL)
+		{
+			return connection;
+		}
+	}
+
+	/*
+	 * For placements that haven't been assigned a connection by a previous command
+	 * in the current transaction, we use a separate connection per placement for
+	 * hash-distributed tables in order to get the maximum performance.
+	 */
+	if (placement->partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		connectionFlags |= CONNECTION_PER_PLACEMENT;
+	}
+
+	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
+
+	if (PQstatus(connection->pgConn) != CONNECTION_OK)
+	{
+		if (stopOnFailure)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+		else
+		{
+			const bool raiseErrors = true;
+
+			HandleRemoteTransactionConnectionError(connection, raiseErrors);
+
+			return NULL;
+		}
+	}
+
+	/*
+	 * Errors are supposed to cause immediate aborts (i.e. we don't
+	 * want to/can't invalidate placements), mark the connection as
+	 * critical so later errors cause failures.
+	 */
+	MarkRemoteTransactionCritical(connection);
+	ClaimConnectionExclusively(connection);
+	RemoteTransactionBeginIfNecessary(connection);
+
+	return connection;
+}
+
+
+/*
+ * GetWorkerConnectionIfCached finds a connection for the given hostname/port from
+ * the given connectionStateHash. If not found, it returns NULL.
+ */
+static MultiConnection *
+GetWorkerConnectionIfCached(const char *nodeName, int nodePort, HTAB *connectionStateHash)
```

Comment:
I think instead of this trick, we could consider using `GetPlacementListConnection()` without `CONNECTION_PER_PLACEMENT `. In that case, we should get the same connection and have the placement connection associations done. That's how sequential router executor does.

Reply:
Note that the connection is also claimed exclusively, which means GetPlacementListConnection will not return it again.

Reply:
> I think instead of this trick, we could consider using `GetPlacementListConnection()` without `CONNECTION_PER_PLACEMENT `. In that case, we should get the same connection and have the placement connection associations done. That's how sequential router executor does.

Removed this function and used the method you and @marcocitus suggested.